### ISSUE TITLE
Improve Mailpile's responsive rendering

### ISSUE THREAD

**Issue by pchampin** [mailpile/Mailpile#1599]
When trying https://demo.mailpile.is/ in my mobile (or on my laptop and reducing the size of the browser window), I notice that some effort is made to make the application responsive. However, I find it unsusable on my Android smartphone (with Firefox and Chrome alike), because everything is too small.

This is due to the fact that smartphone are lying about their resolution, in order to correctly display non-responsive websites. However, a propertly responsive application should inform the smartphone _not_ to lie about it.

Although the future standard is [not yet supported](http://caniuse.com/#search=%40viewport), the current de-facto solution, namely [&lt;meta name="viewport"&gt;](https://developer.mozilla.org/en-US/docs/Mozilla/Mobile/Viewport_meta_tag?redirectlocale=en-US&redirectslug=Mobile%2FViewport_meta_tag) seems to work in many situations.

This would be a very nice improvement.

**Comment by broder** [comment]
Seems to have been added here: https://github.com/mailpile/Mailpile/blob/74d1eea99bdb892555b0cf191b612b9340df1c48/shared-data/default-theme/html/partials/head.html#L3

**Comment by jamestomasino** [comment]
Some quick testing makes it look like there will be some follow-up work on this change. Once the viewport meta tag is in place, the mobile styles for the default theme are wider than the device.

![mobile-styles](https://cloud.githubusercontent.com/assets/201422/23324417/a8041360-fabb-11e6-9564-4100b69e667e.jpg)

**Comment by BjarniRunar** [comment]
Yes, more work is needed. Renamed the issue to reflect that!

**Comment by grekko** [comment]
@BjarniRunar Do you have any specific goals already on the top of your head?

I shortly checked some pages with a limited viewport. I decided to test it with 320 CSS pixels width (e.g. the iPhone 5 comes with that) which seems to be the lower boundary for most known devices according to https://mydevice.io/devices/

I checked some pages, e.g. the `/profiles/` url and most actions thing seem to be accessible. The layout is a little bit funny in some places but that does not prevent me from accessing the page.

![The `/profiles/`-url with 320 CSS pixels width](https://dl.dropboxusercontent.com/s/q3pfytq6337rdut/2017-04-25%20at%2012.39%20%281%29.png)

## Some things I noticed

**1) The notifications widget** takes a lot of space in comparison to the actual content and *blocks* direct access to the search form. The widget could be hidden behind a smaller Icon that "lights up" if something important happens and clicking it would open the full view.

**2) The page description (?)** is a little bit off. We could just fix the layout (and get rid of all the absolute positioned elements)

**3) The headline and home icon** are maybe a little bit noisy? They take up a lot of space and could be hidden instead. The page description already told me what I can do here.

**4) Responsive Tables are hard** but everything is already mostly accessible. Only some icons might be a little bit small to properly be touched.

**5) The bottom navigation** looks good. The only thing I observed is that I did not immediately understand that I could verticially scroll inside the navigation :)

**Comment by largestprime** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

just to add some observations of my own

**6) on some smaller resolutions, the logout button disappears
(actually the whole panel of top-right buttons disappear). add
them to the navigation somewhere?

**7) the little sliver of clickable-draggable area on the far
left of each message (particularly in list view) sometimes
disappears and sometimes becomes too small to be useable. would
it be possible, as a general thing, to either make that area much
larger or even for the whole message to be draggable?

**8) in list view on narrow resolutions, the 'sender' field
disappears. is it possible to make sender and subject stack on
top of each other when the view is very narrow?

Gregory Igelmund <notifications@github.com> wrote:
> @BjarniRunar Do you have any specific goals already on the top
> of your head?
> 
> I shortly checked some pages with a limited viewport. I decided
> to test it with 320 CSS pixels width (e.g. the iPhone 5 comes
> with that) which seems to be the lower boundary for most known
> devices according to https://mydevice.io/devices/
> 
> I checked some pages, e.g. the `/profiles/` url and most
> actions thing seem to be accessible. The layout is a little bit
> funny in some places but that does not prevent me from
> accessing the page.
> 
> ![The `/profiles/`-url with 320 CSS pixels
> width](https://dl.dropboxusercontent.com/s/q3pfytq6337rdut/2017-04-25%20at%2012.39%20%281%29.png)
> 
> ## Some things I noticed
> 
> **1) The notifications widget** takes a lot of space in
> comparison to the actual content and *blocks* direct access to
> the search form. The widget could be hidden behind a smaller
> Icon that "lights up" if something important happens and
> clicking it would open the full view.
> 
> **2) The page description (?)** is a little bit off. We could
> just fix the layout (and get rid of all the absolute positioned
> elements)
> 
> **3) The headline and home icon** are maybe a little bit noisy?
> They take up a lot of space and could be hidden instead. The
> page description already told me what I can do here.
> 
> **4) Responsive Tables are hard** but everything is already
> mostly accessible. Only some icons might be a little bit small
> to properly be touched.
> 
> **5) The bottom navigation** looks good. The only thing I
> observed is that I did not immediately understand that I could
> verticially scroll inside the navigation :)
>

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1

iF4EAREKAAYFAlkAixUACgkQSVytamcyAuXiEgD/SqalEYqknmPR1/rXcnKPHhCq
41SPNB9TgirhJoVGkMoA/0UInHekORepNHMDPSV6NOS0XmZRRZHw5w24p/+bPmCE
=hlN8
-----END PGP SIGNATURE-----

**Comment by BjarniRunar** [comment]
Hey @largestprime and @grekko. I agree with all of your points above.

Daniel's point 8 and Gregory's point 4 are related. I wanted to implement 8, but because responsive tables are hard I gave up during my last iteration. It would be awesome if someone with better CSS skills than mine could get that to work. Related, I'd add:

9) The lines of the result list aren't tall enough to allow for reliable thumb-navigation IMO. Making them taller (by adding the Sender?) would make things more usable.

**Comment by BjarniRunar** [comment]
Lots of progress in #1868 ... :+1:

**Comment by smari** [comment]
Have done some work on this, specifically on 3, 4 and 8. The other things are nice to have, but not implemented yet.

**Comment by Kabouik** [comment]
As requested by @JocelynDelalande, I repost shere some ideas that I mistakenly posted in a closed PR:

```

After I finally installed Mailpile on a server, I can access it from my smartphone. It's great and I'm really excited about it. However, in my opinion there is still too much UI stuff and not enough content to make it comfortable to use. Seeing just 2.5 truncated rows in Inbox view (above screenshots from @JocelynDelalande) brings a lot of frustration when checking emails.

The above pictures from @JocelynDelalande show it in portrait mode, and the below pictures show it in landscape mode on a 5" screen (this is a Sailfish phone, but I've tried with an Android browser and it's basically the same except for the URL bar being at the top):

![](http://i.imgur.com/DyOYI0f.png)

This is the minimal and default zoom level and I think it's just not enough. Since it's both the minimum and default, the only thing you can do is zoom in, but it's not really useful or relevant: 

![](http://i.imgur.com/7xxWTL0.png)

Overall, I believe there are a few key points where space could be gained because it basically does not bring any improved ergonomy:

![](http://i.imgur.com/YCcYqPQ.png)

Just let me emphasize that these "Too something", "Not enough something" are just opinions and are not meant to be as peremptory as they look, I just wanted to keep them short to fit in the screenshot.

1. Buttons on the left are convenient because they are big and easy to select with thumb, but there is still some significant space loss. Overall, the grey bar could be narrower by reducing margins around icons (and margins between icons and their text). For reference, the buttons in the URL bar of my browser are totally big enough for thumbs; they still have big margins but they don't actually need them, see for instance the URL bar in another native browser with smaller margins (and still no issue with using thumbs on these buttons):
![image](https://user-images.githubusercontent.com/7107523/29291852-39143c4e-8145-11e7-8007-840ae6248d58.png)

2. Rows need to be high enough to be easy to select, but I think it's still too big for what is the minimal zoom level. It shows only a total of 3.5 rows on a 5" screen. Row height could probably be reduced to the height of the contact icons on the left and still be easy to select. The Display snug/cozy/comfy options do not do much on mobile view. This is snug here.

3. The information row about selection checkboxes doesn't seem critical to take this much space. Wouldn't users just realize that they can select emails by taping on them without this information? Do we really need a "Select all" in mobile view? What about removing this row, but moving individual checkboxes onto the corner of contact icons? This would make it obvious that emails can be selected, and this would also indicate where to tap to select (Mailpile doesn't clearly indicate areas where to click to select/open/search by contact). Of course the rows shows extra options when something is selected, like tags, but this could be displayed only when there is an active selection.

4. There is lost space here, at least in landscape mode. This is less obvious in portrait but the grey bar height is still huge (and to a lesser extent, the height of the search field seems a bit big as well). Do we need it to be displayed all the time? What about showing it just when at the top of Mailpile view, but hiding it when scrolling down? 

5. There is lost space as well, and actually the left icons are search icons. What about moving them to the  top bar next to the search field? This would make a consistent "Search" bar with all search tools in the same place, and Order and Display could be grouped together in the top right (and still shown only when at the top of the window, it's not necessary when scrolling down). This would work best in landscape mode, but would be harder to achieve in landscape of course (unless all search icons could be all grouped in a single hamburger icon?).

6. After playing a bit with the search field on mobile view, I noticed it is not wide enough to show a full email address, at least at the current minimal zoom level with the current font size. This means that if you want to edit a search or add search terms, you have to clear it beforehand with the "X" button on the right. You may hit this button anyway if you try to place the cursor at the end of your search terms. A wider field would help, or just smaller font, or a combination of both, with just enough space on the right to place the search icons and order/display menu. In landscape, there's a lot of space for whatever you'd like. In portrait, maybe the bar could just be [Mailpile icon, Search field (without the blue button, who would use it instead of just hitting return on the keyboard?), Search icons grouped in a hamburger icon, Order/Display menu]. This way it would fit even in portrait mode. Probably nobody would use any of these tools on a smartphone when scrolling so it'd be fine if they only show up when at the top of the page.

7. I did not comment it on the screenshots but the notification tab could be less visible or completely hidden when there are no notifications, or when they are dismissed. It could also just be an icon that expands when relevant I guess,

8. Since Sailfish comes with a full-fledged terminal experience, I can ssh into my server and play with Mailpile commands, but can't do it with the UI built-in terminal since there is no way to press T out of a text field. Not all phones have a full terminal, and I suppose the point of the built-in Mailpile terminal is to use it from the UI. Would it be possible to add a terminal button somewhere in the Accounts or Privacy pages of the UI? [Edit] The terminal button is already there, sorry about that. However the built-in terminal does not seem to show as much information as the real one: no welcome message with the regular checks that are done when Mailpile is started, no output on `sendmail` (it fails in my case, but I still can't see it from the built-in terminal).

These are just ideas/suggestions, and I figured I'd just write them down before they go away (I don't think I will use Mailpile mobile view extensively before it can display more content, so I could forget quickly), but of course you do whatever you want with them, including burning and throwing to the trash. I do know that the mobile UI is work in progress and that at the moment there is no real mobile vs desktop views, but rather resolution-dependent layouts, and therefore all the above suggestions would be *a lot* of work to implement, but again this is just a list of improvement ideas you may agree with or not; not demands or urgent requests or anything like that. Most of the issues could be worked around with just allowing a smaller font and lower level of minimal zoom (either as a temporary or permanent solution). In any case, it's already great to see that most functions do work on a mobile browser and I am very grateful that you guys already spent some time on it to make it real!
```

Note that since this post, I have been trying to make my Mailpile more user-friendly for mobile view with a custom css style. I've tried only portrait mode (and in fact I would prefer landscape mode because my next phone will have a hardware keyboard; **please give some love to mobile view in landscape mode too**). 

![2017-08-20_234940](https://user-images.githubusercontent.com/7107523/29498788-6ffb973a-8603-11e7-8102-f6ef2bb15ea4.png)

I think it looks good, however it's totally non functional because:
1. It is very resolution-dependent and the resolution I used in my browser is not even the actual resolution of my phone, I was just trying some things quickly to see if it would be even possible to make the view cleaner)
2. I can't make the search bar *and* the sub-navigation icons next to it have be clickable together. It's either the search bar (and Mailpile icon), or the sub-navigation icons. I have tried fiddling with z-index and so on but no success. I suppose we would need to make the top bar and the sub-navigation bar be only 50% width but I did not know how to do that.
3. It's totally incompatible with the full Mailpile view in a desktop resolution. Again, I was just trying some quick css edits to see if it would be possible to declutter the view and show more content without altering mobile ergonomy with just css changes. Someone who actually knows about css could probably do that properly.

But it still better expresses the ideas I was suggesting in the above post.

**Comment by BjarniRunar** [comment]
I'm going to close this issue - the mobile UI is now usable enough that we should be filing individual issues for specific problems, rather than one big catch-all like this. Thanks for all the input folks, keep it coming! 😀
